lib/fathead/python/redirect.py

```diff
@@ -0,0 +1,170 @@
+        except Exception as e:
+            raise BadEntryException('Article had invalid number of elements.')
+
+        if self.entry_type == 'A':
+            self.parse_alternative_keys()
+
+        return self.data
+
+    def parse_alternative_keys(self):
+        """
+        Find alternative keys to use in generated redirects
+
+        Returns:
+            Set of possible redirect entries
+        """
+        self.alternative_keys = set()
+
+        if '.' in self.key and self.entry_type == 'A':
+            key_arr = self.key.split('.')
+            method_name = key_arr[-1]
+            package_name = key_arr[0]
+            new_key = "{} {}".format(package_name, method_name)
+            self.alternative_keys.add(new_key)
+            self.alternative_keys.add(method_name)
+        return self.alternative_keys
+
+    def get_data(self):
+        return self.data
+
+    def get_key(self):
+        return self.key
+
+    def get_type(self):
+        return self.entry_type
+
+    def get_redirects(self):
+        redirs = []
+
+        for alt_key in list(self.alternative_keys):
+            entry = Entry([
+                alt_key,
+                'R',
+                self.key
+            ])
+            redirs.append(entry)
+
+        return redirs
+
+    def get_alternatives(self):
+        return self.alternative_keys
+
+    def get_entry(self):
+        return '\t'.join([
+            self.key,             # title / key
+            self.entry_type,      # entry type
+            self.reference,       # no redirect data
+            '',                   # ignore
+            '',                   # no categories
+            '',                   # ignore
+            '',                   # no related topics
+            '',                   # ignore
+            '',                   # add an external link back to page
+            '',                   # no disambiguation
+            '',                   # images
+            self.abstract,        # abstract
+            self.anchor           # anchor to specific section
+        ])
+
+    def __str__(self):
+        return self.get_entry()
+
+def generate_redirects(f):
+    # @TODO: Cleanup
+    # @TODO: Figure out why the duplicate count is ~20k
+    output = dict()
+
+    # For debugging purposes
+    duplicate_count = 0
+
+    for line in f.readlines():
+        try:
+            # Parse entry
+            entry = Entry(line)
+            key = entry.get_key()
+
+            # Do we have the entry yet?
+            if key not in output:
+                output[key] = str(entry)
+            else:
+                duplicate_count += 1
+
+            # Get all possible redirect entries
+            redirects = entry.get_redirects()
+            for redirect in redirects:
+                key = redirect.get_key()
+                if key not in output:
+                    output[key] = str(entry)
+                else:
+                    duplicate_count += 1
```

Comment:
I think we want to remove the key from `output` if we see a duplicate? That would mean there is some namespace conflict.